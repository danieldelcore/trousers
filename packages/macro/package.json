{
    "name": "@trousers/macro",
    "version": "3.0.2",
    "main": "dist/trousers-macro.cjs.js",
    "module": "dist/trousers-macro.esm.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {
        "@trousers/core": "3.0.2",
        "babel-plugin-macros": "^3.0.0"
    },
    "devDependencies": {
        "react": "^17.0.1"
    },
    "peerDependencies": {
        "react": ">=16.8.0"
    },
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/macro"
}
