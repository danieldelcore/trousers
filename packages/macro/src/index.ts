//@ts-ignore
export * from './macro';
export { css, jsx } from './stub';
