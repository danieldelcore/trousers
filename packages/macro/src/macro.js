const { createMacro } = require('babel-plugin-macros');
const { parse } = require('@babel/parser');
const { process, themify } = require('@trousers/core');
const hash = require('@trousers/hash').default;

const parseObject = objectExpression =>
    objectExpression.properties.reduce((accum, { key, value }) => {
        let parsedValue;

        if (
            value.type === 'StringLiteral' ||
            value.type === 'NumericLiteral' ||
            value.type === 'BooleanLiteral'
        ) {
            parsedValue = value.value;
        } else {
            parsedValue = parseObject(value);
        }

        accum[key.name || key.value] = parsedValue;

        return accum;
    }, {});

function macro({ references, babel }) {
    const t = babel.types;

    if (references.css.length === 0) return;

    references.css.forEach(reference => {
        const styleBlocks = [];
        const importName = reference.node.name;

        reference.find(path => {
            if (!path.isCallExpression()) return;

            if (
                path.node.callee.name === importName ||
                ['modifier', 'theme', 'global'].includes(
                    path.node.callee.property.name,
                )
            ) {
                styleBlocks.push(path);
            }
        });

        let elementId = '';

        styleBlocks.forEach(styleBlock => {
            const { arguments, callee } = styleBlock.node;
            const objectExpression =
                arguments.length === 2 ? arguments[1] : arguments[0];
            const type = callee.name || callee.property.name;
            const rawStyleBlock = parseObject(objectExpression);
            const hashedStyles = hash(JSON.stringify(rawStyleBlock));

            let id = arguments.length === 2 ? arguments[0].value : '';
            let className = '';
            let processedStyles;

            switch (type) {
                case importName:
                    elementId = id;
                    className = `.${id && id + '-'}${hashedStyles}`;
                    processedStyles = process(className, rawStyleBlock);
                    break;
                case 'modifier':
                    className = `.${elementId}--${arguments[0].value}-${hashedStyles}`;
                    processedStyles = process(className, rawStyleBlock);
                    break;
                case 'global':
                    id = `global-${elementId}-${hashedStyles}`;
                    processedStyles = process(className, rawStyleBlock);
                    break;
                case 'theme':
                    className = `.theme-${elementId}-${hashedStyles}`;
                    processedStyles = process(
                        className,
                        themify(rawStyleBlock),
                    );
                    break;
                default:
                    break;
            }

            const parsedStyles = parse(`(${JSON.stringify(processedStyles)})`);

            styleBlock.replaceWith(
                t.callExpression(callee, [
                    t.stringLiteral(id),
                    parsedStyles.program.body[0].expression,
                ]),
            );
        });
    });

    const importName = references.css[0].node.name;

    const { body } = references.css[0].findParent(path =>
        path.isProgram(),
    ).node;

    body.unshift(
        t.importDeclaration(
            [
                t.importSpecifier(
                    t.identifier(importName),
                    t.identifier(importName),
                ),
            ],
            t.stringLiteral('@trousers/macro/runtime'),
        ),
    );

    return {
        keepImports: false,
    };
}

module.exports = createMacro(macro);
