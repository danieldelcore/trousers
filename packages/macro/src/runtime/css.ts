interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: Record<string, string>;
}

export interface CollectorReturn {
    modifier: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    global: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    theme: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: Definition['id'], elementStyles: Definition['styles']) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
