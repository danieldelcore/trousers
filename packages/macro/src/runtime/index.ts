export { default as css } from './css';
export { default as TrousersNested } from './jsx';
