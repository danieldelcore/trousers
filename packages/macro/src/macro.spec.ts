import { transformSync } from '@babel/core';
import plugin from 'babel-plugin-macros';

const transform = (code: TemplateStringsArray) => {
    return transformSync(code[0], {
        configFile: false,
        babelrc: false,
        filename: __filename,
        presets: [['@babel/preset-react', { runtime: 'automatic' }]],
        plugins: [plugin],
    })?.code;
};

describe('macro', () => {
    describe('when parsing css collectors', () => {
        it('should process element collector', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'blue' });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2561700995\\": \\"color: blue;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process element collector without an identifier', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css({ color: 'blue' });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"\\", {
                  \\".2561700995\\": \\"color: blue;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with a single modifier', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'blue' })
                .modifier('primary', { color: 'brown' });
              const App = () => <button css={styles} $primary={true}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2561700995\\": \\"color: blue;\\"
                }).modifier(\\"primary\\", {
                  \\".Button-2561700995--primary-2270159875\\": \\"color: brown;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  $primary: true,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with multiple modifiers', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'blue' })
                .modifier('primary', { color: 'brown' })
                .modifier('secondary', { color: 'purple' });
              const App = ({primary, secondary}) => {
                <button css={styles} $primary={primary} $secondary={secondary}>
                  Submit
                </button>
              };
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2561700995\\": \\"color: blue;\\"
                }).modifier(\\"primary\\", {
                  \\".Button-2561700995--primary-2270159875\\": \\"color: brown;\\"
                }).modifier(\\"secondary\\", {
                  \\".Button-2561700995--secondary-3026956261\\": \\"color: purple;\\"
                });

                const App = ({
                  primary,
                  secondary
                }) => {
                  /*#__PURE__*/
                  _jsx(TrousersNested, {
                    css: styles,
                    $primary: primary,
                    $secondary: secondary,
                    elementType: \\"button\\",
                    children: \\"Submit\\"
                  });
                };"
            `);
        });

        it('should process collector with many modifiers', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'blue' })
                .modifier('primary', { color: 'brown' })
                .modifier('secondary', { color: 'purple' })
                .modifier('tertiary', { color: 'yellow' })
                .modifier('quaternary', { color: 'green' });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2561700995\\": \\"color: blue;\\"
                }).modifier(\\"primary\\", {
                  \\".Button-2561700995--primary-2270159875\\": \\"color: brown;\\"
                }).modifier(\\"secondary\\", {
                  \\".Button-2561700995--secondary-3026956261\\": \\"color: purple;\\"
                }).modifier(\\"tertiary\\", {
                  \\".Button-2561700995--tertiary-41860765\\": \\"color: yellow;\\"
                }).modifier(\\"quaternary\\", {
                  \\".Button-2561700995--quaternary-2402939536\\": \\"color: green;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with a theme', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'var(--brand-background)' })
                .theme({ brand: { background: 'green' }});
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2140373281\\": \\"color: var(--brand-background);\\"
                }).theme(\\"\\", {
                  \\".theme-Button-2140373281-1537299292\\": \\"--brand-background: green;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with a deeply nested theme', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'var(--brand-background)' })
                .theme({
                  neutral: '#fff',
                  brand: {
                    forground: 'yellow',
                    background: 'green',
                  },
                });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2140373281\\": \\"color: var(--brand-background);\\"
                }).theme(\\"\\", {
                  \\".theme-Button-2140373281-3270977004\\": \\"--neutral: #fff;--brand-forground: yellow;--brand-background: green;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with a global', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', {})
                .global({
                  ':root': {
                    'backgroundColor': 'red',
                  }
                });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {}).global(\\"global-Button-3938-480010618\\", {
                  \\":root\\": \\"background-color: red;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should process collector with an aliased import name', () => {
            const result = transform`
              import { css as foo } from './macro';
              const styles = foo('Button', { color: 'blue' });

              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { foo, TrousersNested } from \\"@trousers/macro/runtime\\";
                const styles = foo(\\"Button\\", {
                  \\".Button-2561700995\\": \\"color: blue;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should do nothing in the case of an unused import', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const App = () => <button>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import React from 'react';

                const App = () => /*#__PURE__*/_jsx(\\"button\\", {
                  children: \\"Submit\\"
                });"
            `);
        });
    });

    describe('when interpolations are detected', () => {
        it('should render elements without interpolations', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 'red' });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-2313942302\\": \\"color: red;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate booleans (BooleanLiteral)', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: true });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-3336976155\\": \\"color: true;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate numbers (NumericLiteral)', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 5 });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-1906181116\\": \\"color: 5;\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate variables (Identifier)', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const styles = css('Button', { color: foo });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    \\"--interpol0\\": foo
                  },
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate functions (CallExpression)', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: foo() });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    \\"--interpol0\\": foo()
                  },
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate evaluations (BinaryExpression)', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const styles = css('Button', { color: 5+5 });
              const App = () => <button css={styles}>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    \\"--interpol0\\": 5 + 5
                  },
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should not add interpolations to jsx element if styles are not in use', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const styles = css('Button', { color: foo });

              const App = () => <button>Submit</button>;
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(\\"button\\", {
                  children: \\"Submit\\"
                });"
            `);
        });

        it('should interpolate styles used by nested elements', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const bar = 'green';
              const styles = css('Button', { color: foo });
              const innerStyles = css('ButtonInner', { color: bar });

              const App = () => (
                <button css={styles}>
                  <span css={innerStyles}>
                    Hello, World!
                  </span>
                </button>
              );
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const bar = 'green';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });
                const innerStyles = css(\\"ButtonInner\\", {
                  \\".ButtonInner-4214944499\\": \\"color: var(--interpol1);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    \\"--interpol0\\": foo
                  },
                  children: /*#__PURE__*/_jsx(TrousersNested, {
                    css: innerStyles,
                    elementType: \\"span\\",
                    style: {
                      \\"--interpol1\\": bar
                    },
                    children: \\"Hello, World!\\"
                  })
                });"
            `);
        });

        it('should interpolate styles used by sibling elements', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const bar = 'green';
              const styles = css('Button', { color: foo });
              const siblingStyles = css('ButtonInner', { color: bar });

              const App = () => (
                <div>
                  <span css={siblingStyles}>
                    Hello, World!
                  </span>
                  <button css={styles}>
                    Submit
                  </button>
                </div>
              );
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { jsxs as _jsxs } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const bar = 'green';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });
                const siblingStyles = css(\\"ButtonInner\\", {
                  \\".ButtonInner-4214944499\\": \\"color: var(--interpol1);\\"
                });

                const App = () => /*#__PURE__*/_jsxs(\\"div\\", {
                  children: [/*#__PURE__*/_jsx(TrousersNested, {
                    css: siblingStyles,
                    elementType: \\"span\\",
                    style: {
                      \\"--interpol1\\": bar
                    },
                    children: \\"Hello, World!\\"
                  }), /*#__PURE__*/_jsx(TrousersNested, {
                    css: styles,
                    elementType: \\"button\\",
                    style: {
                      \\"--interpol0\\": foo
                    },
                    children: \\"Submit\\"
                  })]
                });"
            `);
        });

        it('should interpolate reused styles', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const styles = css('Button', { color: foo });

              const App = () => (
                <button css={styles}>
                  <span css={styles}>
                    Hello, World!
                  </span>
                </button>
              );
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    \\"--interpol0\\": foo
                  },
                  children: /*#__PURE__*/_jsx(TrousersNested, {
                    css: styles,
                    elementType: \\"span\\",
                    style: {
                      \\"--interpol0\\": foo
                    },
                    children: \\"Hello, World!\\"
                  })
                });"
            `);
        });

        it('should add interpolations to an in-use style attribute', () => {
            const result = transform`
              import React from 'react';
              import { css } from './macro';
              const foo = 'blue';
              const styles = css('Button', { color: foo });

              const App = () => (
                <button css={styles} styles={{color: 'red'}}>
                    Hello, World!
                </button>
              );
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React from 'react';
                const foo = 'blue';
                const styles = css(\\"Button\\", {
                  \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                });

                const App = () => /*#__PURE__*/_jsx(TrousersNested, {
                  css: styles,
                  elementType: \\"button\\",
                  style: {
                    color: 'red',
                    \\"--interpol0\\": foo
                  },
                  children: \\"Hello, World!\\"
                });"
            `);
        });

        it('should interpolate styles passed directly into the css prop', () => {
            const result = transform`
              import React, { useState } from 'react';
              import { css } from './macro';

              const App = () => {
                const [foo, setFoo] = useState('blue');

                return (
                  <button css={css('Button', { color: foo })}>
                      Hello, World!
                  </button>
                );
              }
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React, { useState } from 'react';

                const App = () => {
                  const [foo, setFoo] = useState('blue');
                  return /*#__PURE__*/_jsx(TrousersNested, {
                    css: css(\\"Button\\", {
                      \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                    }),
                    elementType: \\"button\\",
                    style: {
                      \\"--interpol0\\": foo
                    },
                    children: \\"Hello, World!\\"
                  });
                };"
            `);
        });

        it('should interpolate styles passed directly into multiple css props', () => {
            const result = transform`
              import React, { useState } from 'react';
              import { css } from './macro';

              const App = () => {
                const [foo, setFoo] = useState('blue');
                const [bar, setBar] = useState('red');

                return (
                  <button css={css('Button', { color: foo })}>
                    <span css={css('Span', { color: bar })}>
                      Hello, World!
                    </span>
                  </button>
                );
              }
            `;

            expect(result).toMatchInlineSnapshot(`
                "import { jsx as _jsx } from \\"react/jsx-runtime\\";
                import { css, TrousersNested } from \\"@trousers/macro/runtime\\";
                import React, { useState } from 'react';

                const App = () => {
                  const [foo, setFoo] = useState('blue');
                  const [bar, setBar] = useState('red');
                  return /*#__PURE__*/_jsx(TrousersNested, {
                    css: css(\\"Button\\", {
                      \\".Button-4214914708\\": \\"color: var(--interpol0);\\"
                    }),
                    elementType: \\"button\\",
                    style: {
                      \\"--interpol0\\": foo
                    },
                    children: /*#__PURE__*/_jsx(TrousersNested, {
                      css: css(\\"Span\\", {
                        \\".Span-4214944499\\": \\"color: var(--interpol1);\\"
                      }),
                      elementType: \\"span\\",
                      style: {
                        \\"--interpol1\\": bar
                      },
                      children: \\"Hello, World!\\"
                    })
                  });
                };"
            `);
        });
    });
});
