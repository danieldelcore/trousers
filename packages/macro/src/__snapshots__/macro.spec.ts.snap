// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@trousers/macro complex theme: complex theme 1`] = `

import { css } from './macro';
css('Button', { color: 'var(--brand-background)' })
  .theme({
    neutral: '#fff',
    brand: {
      forground: 'yellow',
      background: 'green',
    },
  });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2140373281': 'color: var(--brand-background);',
}).theme('', {
    '.theme-Button-3270977004':
        '--neutral: #fff;--brand-forground: yellow;--brand-background: green;',
});


`;

exports[`@trousers/macro correctly interpolates booleans (BooleanLiteral): correctly interpolates booleans (BooleanLiteral) 1`] = `

import { css } from './macro';
css('Button', { color: true });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-3336976155': 'color: true;',
});


`;

exports[`@trousers/macro correctly interpolates booleans (NumericLiteral): correctly interpolates booleans (NumericLiteral) 1`] = `

import { css } from './macro';
css('Button', { color: 5 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-1906181116': 'color: 5;',
});


`;

exports[`@trousers/macro correctly updates imports: correctly updates imports 1`] = `

/** @jsx jsx */
import { css, jsx } from './macro';

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`@trousers/macro correctly updates renamed imports: correctly updates renamed imports 1`] = `

/** @jsx bar */
import { css as foo, jsx as bar } from './macro';

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`@trousers/macro different import name: different import name 1`] = `

import { css as foo } from './macro';
foo('Button', { color: 'blue' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { foo } from '@trousers/macro/runtime';
foo('Button', {
    '.Button-2561700995': 'color: blue;',
});


`;

exports[`@trousers/macro element without identifier: element without identifier 1`] = `

import { css } from './macro';
css({ color: 'blue' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('', {
    '.2561700995': 'color: blue;',
});


`;

exports[`@trousers/macro element: element 1`] = `

import { css } from './macro';
css('Button', { color: 'blue' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2561700995': 'color: blue;',
});


`;

exports[`@trousers/macro global: global 1`] = `

import { css } from './macro';
css('Button', {})
  .global({
    ':root': {
      'backgroundColor': 'red',
    }
  });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {}).global('global-Button-480010618', {
    ':root': 'background-color: red;',
});


`;

exports[`@trousers/macro many modifiers: many modifiers 1`] = `

import { css } from './macro';
css('Button', { color: 'blue' })
  .modifier('primary', { color: 'brown' })
  .modifier('secondary', { color: 'purple' })
  .modifier('tertiary', { color: 'yellow' })
  .modifier('quaternary', { color: 'green' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2561700995': 'color: blue;',
})
    .modifier('primary', {
        '.Button--primary-2270159875': 'color: brown;',
    })
    .modifier('secondary', {
        '.Button--secondary-3026956261': 'color: purple;',
    })
    .modifier('tertiary', {
        '.Button--tertiary-41860765': 'color: yellow;',
    })
    .modifier('quaternary', {
        '.Button--quaternary-2402939536': 'color: green;',
    });


`;

exports[`@trousers/macro modifiers: modifiers 1`] = `

import { css } from './macro';
css('Button', { color: 'blue' })
  .modifier('primary', { color: 'brown' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2561700995': 'color: blue;',
}).modifier('primary', {
    '.Button--primary-2270159875': 'color: brown;',
});


`;

exports[`@trousers/macro multiple modifiers: multiple modifiers 1`] = `

import { css } from './macro';
css('Button', { color: 'blue' })
  .modifier('primary', { color: 'brown' })
  .modifier('secondary', { color: 'purple' });

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2561700995': 'color: blue;',
})
    .modifier('primary', {
        '.Button--primary-2270159875': 'color: brown;',
    })
    .modifier('secondary', {
        '.Button--secondary-3026956261': 'color: purple;',
    });


`;

exports[`@trousers/macro theme: theme 1`] = `

import { css } from './macro';
css('Button', { color: 'var(--brand-background)' })
  .theme({ brand: { background: 'green' }});

      ↓ ↓ ↓ ↓ ↓ ↓

import { css } from '@trousers/macro/runtime';
css('Button', {
    '.Button-2140373281': 'color: var(--brand-background);',
}).theme('', {
    '.theme-Button-1537299292': '--brand-background: green;',
});


`;

exports[`@trousers/macro unused import: unused import 1`] = `

import { css } from './macro';

      ↓ ↓ ↓ ↓ ↓ ↓



`;
